pkg/coscheduling/coscheduling.go

```diff
@@ -0,0 +1,279 @@
+	}
+	if pgInfo2 != nil {
+		time2 = pgInfo2.InitialTimestamp
+	}
+
+	if pgInfo1 != nil && pgInfo2 != nil && time1.Equal(time2) {
+		return pgInfo1.ID < pgInfo2.ID
+	}
+	return time1.Before(time2)
+}
+
+func (cs *Coscheduling) getPodGroupInfo(p *v1.Pod) *PodGroupInfo {
+	podGroupName, min := GetPodGroupLabels(p)
+	if podGroupName != "" {
+		key := fmt.Sprintf("%v/%v", p.Namespace, podGroupName)
+		pgInfo, ok := cs.PodGroupInfos[key]
+		if !ok {
+			pgInfo = PodGroupInfo{
+				Name:             podGroupName,
+				InitialTimestamp: p.CreationTimestamp.Time,
+				ID:               cs.GetID(),
+				MinAvailable:     min,
+			}
+			cs.PodGroupInfos[key] = pgInfo
+		}
+		return &pgInfo
+	}
+	return nil
+}
+
+func (cs *Coscheduling) PreFilter(ctx context.Context, state *framework.CycleState, p *v1.Pod) *framework.Status {
+	podGroupName, minAvailable := GetPodGroupLabels(p)
+	if podGroupName == "" || minAvailable <= 1 {
+		return framework.NewStatus(framework.Success, "")
+	}
+
+	total := cs.calculateTotalPods(podGroupName, p.Namespace)
+	if total < minAvailable {
+		klog.V(3).Infof("The count of podGroup %v/%v is not up to minAvailable(%d) in PreFilter: %d",
+			p.Namespace, podGroupName, minAvailable, total)
+		return framework.NewStatus(framework.Unschedulable, "less than minAvailable")
+	}
+
+	return framework.NewStatus(framework.Success, "")
+}
+
+func (cs *Coscheduling) PreFilterExtensions() framework.PreFilterExtensions {
+	return nil
+}
+
+// Permit is the functions invoked by the framework at "permit" extension point.
+func (cs *Coscheduling) Permit(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) (*framework.Status, time.Duration) {
+	podGroupName, minAvailable := GetPodGroupLabels(p)
+	if podGroupName == "" || minAvailable <= 1 {
+		return framework.NewStatus(framework.Success, ""), 0
+	}
+
+	namespace := p.Namespace
+	running := cs.calculateRunningPods(podGroupName, namespace)
+	waiting := cs.calculateWaitingPods(podGroupName, namespace)
+	current := running + waiting + 1
+
+	if current < minAvailable {
+		klog.V(3).Infof("The count of podGroup %v/%v is not up to minAvailable(%d) in Permit: running(%d), waiting(%d)",
+			namespace, podGroupName, minAvailable, running, waiting)
+		// TODO Change the timeout to dynamic value depends on the size of the `PodGroup`
+		return framework.NewStatus(framework.Wait, ""), 10 * PermitWaitingTime
+	}
+
+	klog.V(3).Infof("The count of podGroup %v is up to minAvailable(%d) in Permit: running(%d), waiting(%d)",
+		podGroupName, minAvailable, running, waiting)
+	cs.FrameworkHandle.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
+		if waitingPod.GetPod().Namespace == namespace && waitingPod.GetPod().Labels[PodGroupName] == podGroupName {
+			klog.V(3).Infof("Permit allows the pod: %v/%v", podGroupName, waitingPod.GetPod().Name)
+			waitingPod.Allow(cs.Name())
+		}
+	})
+
+	delete(cs.PodGroupInfos, fmt.Sprintf("%v/%v", namespace, podGroupName))
```

Comment:
By deleting the entry here, the next Pod with the same podGroupName will form a new PodGroupInfo with a new ID.   is this intended? 

Maybe, we should implement a timeout based gc for the PodGroupInfos map.

Reply:
SGTM, I add to TODO.